yew/src/lib.rs

```diff
@@ -99,7 +99,119 @@ extern crate self as yew;
 /// This macro implements JSX-like templates.
 pub use yew_macro::html;
 
-#[doc(hidden)]
+/// This macro is similar to [`html!`], but returns actual component type instead
+/// of generic [`Html`].
+///
+/// That macro is useful when, for example, in typical implementation of `List`
+/// component. In the typical implementation you have two connected components --
+/// `List` and `ListItem`, and `List` allowing only `ListItem`s inside them.
+/// Example of such component you can find in the [`nested_list`] example.
+///
+/// Manually you create list by just nesting `ListItem`'s into `List`:
+///
+/// ```
+/// # use yew::prelude::*;
+/// use yew::html::ChildrenRenderer;
+/// use yew::virtual_dom::VChild;
+///
+/// #[derive(Clone, Properties)]
+/// struct List {
+///   children: ChildrenRenderer<ListItem>,
+/// }
+/// impl Component for List {
+///   // ...
+/// #   type Message = ();
+/// #   type Properties = Self;
+/// #   fn create(props: Self::Properties, _: ComponentLink<Self>) -> Self { props }
+/// #   fn update(&mut self, _: Self::Message) -> ShouldRender { false }
+/// #   fn change(&mut self, _: Self::Properties) -> ShouldRender { false }
+/// #   fn view(&self) -> Html { unimplemented!() }
+/// }
+///
+/// #[derive(Clone)]
+/// struct ListItem;
+/// impl Component for ListItem {
+/// #   type Message = ();
+///   type Properties = ();
+///   // ...
+/// #   fn create(_: Self::Properties, _: ComponentLink<Self>) -> Self { Self }
+/// #   fn update(&mut self, _: Self::Message) -> ShouldRender { false }
+/// #   fn change(&mut self, _: Self::Properties) -> ShouldRender { false }
+/// #   fn view(&self) -> Html { unimplemented!() }
+/// }
+///
+/// // Required for ChildrenRenderer
+/// impl From<VChild<ListItem>> for ListItem {
+///   fn from(child: VChild<ListItem>) -> Self { Self }
+/// }
+///
+/// impl From<ListItem> for Html {
+///   fn from(item: ListItem) -> Html { item.view() }
+/// }
```

Comment:
Won't be able to implement a trait on `Html` outside of the crate.

```suggestion
/// impl Into<Html> for ListItem {
///   fn into(self) -> Html { item.view() }
/// }
```

Reply:
No, it is possible, I checked that. `From` has generic argument, defined in own crate, so it is possible. Also rust guidlines advises to implement `From` rather that `Into`. In addition, documentation tests uses its own crate, and as you can see, they are compiled

Reply:
You're right, since Rust 1.41 it's possible to implement parametrized traits on foreign types. I forgot about that, sorry.

There is however a precedent to use `Into<Html>` in other places like in the `nested_list` example:
https://github.com/yewstack/yew/blob/7301abea14334a3248fd658fc4473167b4e4ec4d/examples/nested_list/src/list.rs#L128

The [official documentation for the `From` trait](https://doc.rust-lang.org/std/convert/trait.From.html) also still states the following:
> Only implement `Into` if a conversion to a type outside the current crate is required. `From` cannot do these type of conversions because of Rust's orphaning rules. See `Into` for more details.

I'm not strictly against using the `From` trait but for the sake of consistency that should be done in a separate PR which updates all the material.

Also, my personal opinion is that implementing `Into` makes more sense semantically in this case.